src/status_im/chat/models/delete_message.cljs

```diff
@@ -0,0 +1,90 @@
+    (update-in db
+               [:messages chat-id message-id]
+               assoc
+               :deleted? true
+               :deleted-undoable-till (+ (datetime/timestamp)
+                                         undo-time-limit-ms))))
+
+(defn- update-db-undo-locally
+  "Restore deleted message if called within timelimit"
+  [db chat-id message-id]
+  (let [{:keys [deleted? deleted-undoable-till]}
+        (get-in db [:messages chat-id message-id])]
+    (if (and deleted?
+             (> deleted-undoable-till (datetime/timestamp)))
+      (update-in db
+                 [:messages chat-id message-id]
+                 dissoc
+                 :deleted?
+                 :deleted-undoable-till)
+      (update-db-clear-undo-timer db chat-id message-id))))
+
+(fx/defn delete
+  "Delete message now locally and broadcast after undo time limit timeout"
+  {:events [:chat.ui/delete-message]}
+  [{:keys [db]} {:keys [chat-id message-id]} undo-time-limit-ms]
+  (when (get-in db [:messages chat-id message-id])
+    (assoc
+     (message-list/rebuild-message-list
+      {:db (update-db-delete-locally db chat-id message-id undo-time-limit-ms)}
+      chat-id)
+     :utils/dispatch-later [{:dispatch [:chat.ui/delete-message-and-send
+                                        {:chat-id    chat-id
+                                         :message-id message-id}]
+                             :ms       undo-time-limit-ms}])))
+
+(fx/defn undo
+  {:events [:chat.ui/undo-delete-message]}
+  [{:keys [db]} {:keys [chat-id message-id]}]
+  (when (get-in db [:messages chat-id message-id])
+    (message-list/rebuild-message-list
+     {:db (update-db-undo-locally db chat-id message-id)}
+     chat-id)))
+
+(fx/defn delete-and-send
+  {:events [:chat.ui/delete-message-and-send]}
+  [{:keys [db]} {:keys [message-id chat-id]}]
+  (when (get-in db [:messages chat-id message-id])
+    {:db             (update-db-clear-undo-timer db chat-id message-id)
+     ::json-rpc/call [{:method      "wakuext_deleteMessageAndSend"
+                       :params      [message-id]
+                       :js-response true
+                       :on-error    #(log/error "failed to delete message " %)
+                       :on-success  #(re-frame/dispatch [:sanitize-messages-and-process-response
+                                                         %])}]}))
+(defn- chats-reducer
+  "traverse all messages find not yet synced deleted? messages, generate dispatch vector"
+  [acc chat-id messages]
+  (reduce-kv
+   (fn [inner-acc message-id {:keys [deleted? deleted-undoable-till]}]
+     (if (and deleted? deleted-undoable-till)
+       (conj inner-acc [:chat.ui/delete-message-and-send chat-id message-id])
+       inner-acc))
+   acc
+   messages))
+
+(fx/defn send-all
+  "Get all deleted messages that not yet synced with status-go and send them"
+  {:events [:chat.ui/send-all-deleted-messages]}
+  [{:keys [db]}]
+  {:dispatch-n (reduce-kv chats-reducer [] (:messages db))})
```

Comment:
I'm sharing this snippet just in case you like it because I found the reduce-kv with another inner reduce-kv call a bit difficult to understand. When I see a reduce call with a filtering step, I think: _Can we decompose it as a `filter` + `map` and would that improve readability? For the trained Clojure eye, a `filter + map` tells a story without even having to scan the code, whereas the reducer function requires a more thorough scan. Please, forgive me if I got your implementation wrong ;)

```clojure
(fx/defn send-all
  {:events [:chat.ui/send-all-deleted-messages]}
  [{:keys [db]}]
  {:dispatch-n (->> db
                    :messages
                    vals
                    (filter #(and (:deleted? %) (:deleted-undoable-till %)))
                    (mapv #(vector :chat.ui/delete-message-and-send chat-id (:id %))))})
```

Reply:
Done